### PR TITLE
nsm driver: add hwrng capabilities

### PR DESCRIPTION
**Pull request by petreeftime**
To allow the NitroSecureModule to seed entropy automatically, add the
capability in the driver to request and send entropy automatically to
the Linux pool by using the hwrng mechanism in Linux.

Note: this requires a kernel with CONFIG_HW_RANDOM enabled.

Signed-off-by: Petre Eftime <epetre@amazon.com>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: nsm-driver/nsm.c [hunk 1]
```diff
@@ -458,6 +459,149 @@ static struct miscdevice nsm_driver_miscdevice = {
...
+		resp_ptr += 9;
+		resp_len -= 9;
+	} else {
+		pr_err("NSM RNG: Invalid CBOR encoding\n");
+		rc = -EFAULT;
+		goto out;
+	}
+
+	/* Return min(resp_len, rand_len, max, INT_MAX) bytes, so rc doesn't overflow */
+	if(resp_len < rand_len) {
```
Comment (alcioa) [inline nsm-driver/nsm.c:583]: This PR needs to be ran against linux checkpatch.

File: nsm-driver/nsm.c [hunk 2]
```diff
@@ -458,6 +459,149 @@ static struct miscdevice nsm_driver_miscdevice = {
...
+	uint64_t resp_len = 0;
+	uint64_t rand_len = 0;
+
+	struct nsm_kernel_message message;
+
+	/*
+	 * 69                          # text(9)
+	 * 47657452616E646F6D          # "GetRandom"
+	 */
+	uint8_t request[] = { 0x69, 0x47, 0x65, 0x74, 0x52, 0x61, 0x6E, 0x64, 0x6F, 0x6D };
```
Comment (alcioa) [inline nsm-driver/nsm.c:475]: Can this be `const uint8_t`?

File: nsm-driver/nsm.c [hunk 3]
```diff
@@ -458,6 +459,149 @@ static struct miscdevice nsm_driver_miscdevice = {
...
+	/*
+	 * A1                          # map(1)
+	 *     69                      # text(9) - Name of field
+	 *         47657452616E646F6D  # "GetRandom"
+	 * A1                          # map(1) - The field itself
+	 *     66                      # text(6)
+	 *         72616E646F6D        # "random"
+	 *	# The rest of the response should be a byte array
+	 */
+	uint8_t response[] = { 0xA1, 0x69, 0x47, 0x65, 0x74, 0x52, 0x61, 0x6E, 0x64,
```
Comment (alcioa) [inline nsm-driver/nsm.c:486]: Can this be `const uint8_t`?

File: nsm-driver/nsm.c [hunk 4]
```diff
@@ -458,6 +459,149 @@ static struct miscdevice nsm_driver_miscdevice = {
...
+	message.request.iov_base = kmalloc(message.request.iov_len, GFP_KERNEL);
+	if (message.request.iov_base == NULL)
+		goto out;
+	memcpy(message.request.iov_base, request, sizeof(request));
+
+	/* Allocate space for response */
+	message.response.iov_len = NSM_RESPONSE_MAX_SIZE;
+	message.response.iov_base = kmalloc(message.response.iov_len, GFP_KERNEL);
+	if (message.response.iov_base == NULL)
+		goto out;
```
Comment (alcioa) [inline nsm-driver/nsm.c:510]: The `message.request.iov_base` will leak here or does `message_delete()` check pointer and free it inside? As well as other `goto out` after this?

File: nsm-driver/nsm.c [hunk 5]
```diff
@@ -458,6 +459,149 @@ static struct miscdevice nsm_driver_miscdevice = {
...
+	if (memcmp(resp_ptr, response, sizeof(response)) != 0) {
+		pr_err("NSM RNG: Invalid response header: Possible error message or invalid response");
+		rc = -EFAULT;
+		goto out;
+	}
+
+	resp_ptr += sizeof(response);
+	resp_len -= sizeof(response);
+
+	if (resp_ptr[0] >> 5 != 0x2) {
```
Comment (alcioa) [inline nsm-driver/nsm.c:536]: I think we should mention somewhere or link the meaning of these magic numbers.

File: nsm-driver/nsm.c [hunk 6]
```diff
@@ -467,26 +611,43 @@ static int __init nsm_driver_init(void)
...
+	hwrng_unregister(&nsm_hwrng);
+err_unreg_misc:
+	misc_deregister(&nsm_driver_miscdevice);
+	nsm_dev = NULL;
 	return rc;
 }
 
 static void __exit nsm_driver_exit(void)
 {
+	hwrng_unregister(&nsm_hwrng);
```
Comment (alcioa) [inline nsm-driver/nsm.c:650]: Does `rmmod` work now?

File: nsm-driver/nsm.c (hunk 4 above)
Comment (petreeftime) [reply nsm-driver/nsm.c:510]: message_delete cleans it up completely.

**Review by popegeo — APPROVED** [review]

**Review by alcioa — APPROVED** [review]

**Comment by jethrogb** [comment]
Great, thanks! I was doing code review and getting very confused as to why things weren't working. Should've checked the source code version against the binary version :upside_down_face:

**Pull request merged**